### PR TITLE
Ability for users to disable default matchers

### PR DESCRIPTION
**Pull request by onetwopunch**
- Added config option to list matchers a users wants to disable
- Created Matcher class with `id` method
- Pass `config` from Scanner entry point down to instantiated classed for better single responsibility.

### CODE REVIEW

File: lib/sanctify/matcher.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      # NOTE: Allow users to specify entire description or something like it.
+      # For example, this would allow a user to disable matchers with description:
+      #
+      disabled_matchers = config['disabled_matchers'] || []
+      disabled_matchers.each do |dis|
+        if dis == id
+          return true
+        end
+      end
+      return false
```
Comment (outreach-devin-canterberry) [inline lib/sanctify/matcher.rb]: L19-24 can be simplified to `disabled_matchers.include?(id)`

File: lib/sanctify/matcher.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+module Sanctify
+  class Matcher
+    attr_reader :description, :regex, :config
+    def initialize(description, regex, config)
```
Comment (outreach-devin-canterberry) [inline lib/sanctify/matcher.rb]: It seems a little odd for an individual matcher to have awareness of the whole gem config (and also expose it via an attribute reader). Considering that all we're using this config for is to discern whether the matcher is disabled, it'd be a tad cleaner to just inject a `disabled:` attribute here.

A usability note: when a method's parameters aren't inherently obvious (almost always true when a method has more than one parameter), it can help to use named parameters instead of positional ones.

```ruby
# A. Positional with options
Sanctify::Matcher.new('Hello, world!', /^Hello, world!$/, disabled: true)

# B. All named parameters
Sanctify::Matcher.new(description: 'Hello, world!', regex: /^Hello, world!$/, disabled: true)
```

🚲 🏚  In comparing the two here, option A seems to make sense and is basically what you have here, with the only exception being that the last argument is an options hash instead of the global'ish gem config.

File: lib/sanctify/matcher.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      @description = description
+      @regex = regex
+      @config = config
+    end
+
+    def id
+      description.downcase.gsub(' ', '_')
+    end
+
+    def disabled?
```
Comment (outreach-devin-canterberry) [inline lib/sanctify/matcher.rb]: In addition to a `#disabled?` method, it might be nice to an `#enabled?` method as well.

File: lib/sanctify/matcher.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  class Matcher
+    attr_reader :description, :regex, :config
+    def initialize(description, regex, config)
+      @description = description
+      @regex = regex
+      @config = config
+    end
+
+    def id
+      description.downcase.gsub(' ', '_')
```
Comment (outreach-devin-canterberry) [inline lib/sanctify/matcher.rb]: This feels like meta-programming, and can make the code hard to follow. Instead, I'd suggest that each matcher define a unique ID for itself **and** a human-readable description, and not couple them like this. As a bonus, it would simplify the logic for determining whether a matcher should be disabled, and allow it to be hoisted to the caller initializing a Matcher object.

With this change, `description` could be moved to the options hash of the constructor, defaulting to the `id`. Something like 

```ruby
def initialize(id, regex, description: nil, disabled: false)
  @id = id
  @regex = regex
  @description = description || id
  @disabled = disabled
end
```

File: lib/sanctify/matcher_list.rb [hunk 5]
```diff
@@ -1,42 +1,64 @@
...
+    def initialize(config)
+      @matchers = DEFAULT_MATCHERS.map{ |obj| Matcher.new(obj[:description], obj[:regex], config) }
+      @config = config
+      initialize_custom_matchers!
     end
 
-    def add(desc:, regex:)
-      if desc.length.zero?
+    def add(description, regex)
+      if description.length.zero?
```
Comment (outreach-devin-canterberry) [inline lib/sanctify/matcher_list.rb]: `description.length.zero?` can be shortened to just `description.empty?`

File: lib/sanctify/matcher_list.rb [hunk 6]
```diff
@@ -1,42 +1,64 @@
...
+        # out a regex that will match base64 but not paths.
         regex: /\b(?<![A-Za-z0-9\/+=])(?=.*[&?=-@#$%\\^+])[A-Za-z0-9\/+=]{40}(?![A-Za-z0-9\/+=])\b/
       },
       {
         description: "SSH RSA Private Key",
         regex: /^-----BEGIN RSA PRIVATE KEY-----$/
       },
       {
-        description: "X.509 Certificate",
+        description: "X509 Certificate",
```
Comment (outreach-devin-canterberry) [inline lib/sanctify/matcher_list.rb]: Was this renamed in order to avoid having dots in the derived matcher `id`? (X.509 was technically correct)
If each matcher had an explicit `id`, we could leave this description be.

File: lib/sanctify/repo.rb [hunk 7]
```diff
@@ -46,5 +46,10 @@ def should_ignore?(path)
     def added_line?(line)
       line.start_with?('+') && !line.start_with?('+++')
     end
+
+    def ignored_paths
+      patterns = config['ignored_paths'] || []
+      patterns.map { |patt| Regexp.new patt }
```
Comment (outreach-devin-canterberry) [inline lib/sanctify/repo.rb]: 🚲 🏚 Not abbreviating `patterns`, but abbreviating `patt`.

File: lib/sanctify/scanner.rb [hunk 8]
```diff
@@ -7,36 +7,16 @@ class Scanner
...
-      custom_matchers = config['custom_matchers'] || []
-      if custom_matchers.any?
-        custom_matchers.each do |cust|
-          if cust['description'] && cust['regex']
-            matcher_list.add(desc: cust['description'], regex: Regexp.new(cust['regex']))
-          else
-            raise ScannerError, "Improperly configured custom matcher: #{cust}. Must include 'description' and 'regex'"
+          next if matcher.disabled?
+          if matcher.regex.match(line)
+            raise ScannerError, "[ERROR] SECRET FOUND (#{matcher.description}): #{line} : #{path}"
```
Comment (outreach-devin-canterberry) [inline lib/sanctify/scanner.rb]: :+1: This (L17-19) reads really nicely!

**Comment by outreach-devin-canterberry** [comment]
First review pass is complete. Some non-blocking bike-sheddy comments, feel free to ignore. I'm digging the extraction of matcher logic into its own class, and the feature of disabling default matchers. This feature makes for a good workaround in the event that there is an issue with one of the default matchers, or a repo using this gem decides that, for whatever reason, some of those default patterns are acceptable.

A couple of things I do want to discuss a bit are:

 1. Deriving a unique `id` for a matcher from its description seems error-prone and counter-intuitive. It already seems to be causing some impact/constraints on how matcher descriptions are written. Seems better for a matcher to be explicit on what its `id` should be, and let the description be unconstrained and human-readable.

2. Injecting the global `config` from the entry-point directly down into internal classes is a bit heavy. Each of these internal classes then needs to know how to navigate the global config, and has carte-blanche access to values that aren't directly relevant. Instead, grokking the config seems like a responsibility for the entry-point (or better yet, a `Config` class). The internal classes (like `Matcher`, `MatcherList`, `Repo`, `Scanner`, et al) could be quite specific about what they need to be given in order to be initialized (vs grokking the config and configuring themselves). For example, `Matcher` only cares about whether it is disabled… `MatcherList` only cares about which matchers it includes… `Repo` only cares about which paths it should ignore… `Scanner` needs to know which repo it is scanning (a `Repo` instance) and which matchers it is using (a `MatcherList` instance).

File: lib/sanctify/matcher.rb (hunk 3 above)
Comment (onetwopunch) [reply lib/sanctify/matcher.rb]: I'm happy to add that when there arises a use for it. For now we only call `disabled?` the one time.

File: lib/sanctify/matcher.rb (hunk 4 above)
Comment (onetwopunch) [reply lib/sanctify/matcher.rb]: Yeah I thought about this before I committed. I did add the list of acceptable ids to the README but as this grows, we'll want to make that more explicit. Good call.

**Review by outreach-devin-canterberry — APPROVED** [review]
:+1: LGTM!
